src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Settings/SiteExtensions.cs

```diff
@@ -0,0 +1,41 @@
+using OrchardCore.Entities;
+
+namespace OrchardCore.Settings;
+
+public static class SiteExtensions
```

Comment:
`ISite` is really about the tenant's persistent user-facing configuration. This is not really adequate for it. Also, if you save properties to it, those might be persisted to the DB (the `AdminController` in `Settings` knows about that, but you can save the current `ISite` otherwise too).

Instead, I recommend a new scoped service specifically for this purpose that uses `ShellScope.AddDeferredTask()` to release the shell once after the normal scope is disposed.

Reply:
I changed it to be more global. So you can signal a request to reload the tenant from anywhere with in the http request context. Now all you need is to call `HttpContext.SignalReleaseShellContext()` from anywhere not just a driver. Then using the new `ShellSettingsReleaseFilter` we add a deferred task to reload the tenant when necessary. I am not sure if a ResultFilter is the best way to add the task or not. Let me know if you think of a better way.

Reply:
Why store this in the `HttpContext`, and especially, why use an extension on it? This is not an HTTP thing but an OC one.

Reply:
Extension for ease of use. On the HttpContext because it is handled at the end of the Http request. It could be updated by multiple services in one request. We could use a service specially for this, but I am not sure it's worth adding a service just to be able to do same thing.

Reply:
It feels to me that this is mixing OC internal things with the `HttpContext`, and using it as kind of a message queue. Which arguably works, but it's akin to using a global static dictionary with conventional keys: simple, but the logic and data aren't encapsulated, and the overall behavior is spread across the codebase (in this case, a static class that's conventionally but not logically extending `HttpContext` with behavior, and the controller).

I'd very much have `IShellRestartQueue.QueueRestart()` (not necessarily this name, but you get the idea) called in the drivers, and everything else is handled in its implementation.

Related: https://github.com/OrchardCMS/OrchardCore/pull/15875#discussion_r1581073402.

Reply:
I added a new service `IShellContextReleaseService` which allow us to request a shell release. Update the implementation and the docs.